### DISCUSSION TITLE
Mounted apps must always start with /

### DISCUSSION THREAD

**Discussion by levrik** [fastapi/fastapi#8986] [Questions]
I'm mounting another ASGI application via `mount` but it always redirects from `/another` to `/another/`. I've found a similar issue with sub-routers (https://github.com/tiangolo/fastapi/issues/414) which was solved.

### Example

```Python
from fastapi import FastAPI

app = FastAPI()

another_app = FastAPI()


@another_app.get("")
def root():
    return "alive"

app.mount("/another", another_app)
```

### Description

I'm using `httpie` for this example.

```
$ http localhost:8000/another
HTTP/1.1 307 Temporary Redirect
Transfer-Encoding: chunked
date: Mon, 04 Jan 2021 08:40:02 GMT
location: http://localhost:8000/another/
server: uvicorn
```

As you can see, a redirect to `/another/` is happening.
I'm just using another `FastAPI` app as example here.
In my real world use-case the other application is the `GraphQL` ASGI app from the Ariadne GraphQL framework.
The behavior is the same. I've also verified that the redirect is triggered by FastAPI by setting a breakpoint in the `GraphQL` ASGI app which wasn't hit.
In the particular use-case shown above it's not even possible to hit the registered URL since it's registered on `""` instead of `"/"`. This doesn't apply to my real world use-case but I could imagine that there are other ASGI apps which don't respond on `"/"` but on `""` only which couldn't be mounted in FastAPI at all due to this.
I realize that probably all ASGI apps register their root routes on `"/"` and not `""` so probably everything is working by design here but in my case I'm getting a lot of redirects from `/graphql` to `/graphql/` now which actually crashes the client app which doesn't handle redirects for security reasons.

### Environment

* OS: [e.g. Linux / Windows / macOS]: macOS
* FastAPI Version [e.g. 0.3.0]: 0.63.0

* Python version: 3.8

**Comment by dstlny** [comment]
> I'm mounting another ASGI application via `mount` but it always redirects from `/another` to `/another/`. I've found a similar issue with sub-routers (#414) which was solved.
> 
> ### Example
> ```python
> from fastapi import FastAPI
> 
> app = FastAPI()
> 
> another_app = FastAPI()
> 
> 
> @another_app.get("")
> def root():
>     return "alive"
> 
> app.mount("/another", another_app)
> ```
> 
> ### Description
> I'm using `httpie` for this example.
> 
> ```
> $ http localhost:8000/another
> HTTP/1.1 307 Temporary Redirect
> Transfer-Encoding: chunked
> date: Mon, 04 Jan 2021 08:40:02 GMT
> location: http://localhost:8000/another/
> server: uvicorn
> ```
> 
> As you can see, a redirect to `/another/` is happening.
> I'm just using another `FastAPI` app as example here.
> In my real world use-case the other application is the `GraphQL` ASGI app from the Ariadne GraphQL framework.
> The behavior is the same. I've also verified that the redirect is triggered by FastAPI by setting a breakpoint in the `GraphQL` ASGI app which wasn't hit.
> In the particular use-case shown above it's not even possible to hit the registered URL since it's registered on `""` instead of `"/"`. This doesn't apply to my real world use-case but I could imagine that there are other ASGI apps which don't respond on `"/"` but on `""` only which couldn't be mounted in FastAPI at all due to this.
> I realize that probably all ASGI apps register their root routes on `"/"` and not `""` so probably everything is working by design here but in my case I'm getting a lot of redirects from `/graphql` to `/graphql/` now which actually crashes the client app which doesn't handle redirects for security reasons.
> 
> ### Environment
> * OS: [e.g. Linux / Windows / macOS]: macOS
> * FastAPI Version [e.g. 0.3.0]: 0.63.0
> * Python version: 3.8

As disgusting as this might seem, the following might be a temporary solution:

 ### Example
```python
from fastapi import FastAPI
app = FastAPI()

if app.router:
   app.router.redirect_slashes = False
```

The only downside with this temporary solution, is all this same behaviour will be applied to all the routes in the app

**Comment by levrik** [comment]
@dstlny This removes the redirect here but sadly I'm getting a `404` then instead. The code of Ariadne framework isn't hit as verified using breakpoints so this response is coming from FastAPI/Starlette.

**Comment by cgorski-airtime** [comment]
I'm getting bit by this is well - anyone have any other ideas at this point?

**Comment by cgorski** [comment]
Creating some middleware worked for me:

```python
from starlette.requests import Request
from starlette.types import ASGIApp, Receive, Scope, Send

class GraphQLRedirect:
    def __init__(
        self,
        app: ASGIApp
    ) -> None:
        self.app = app
        
    async def __call__(self, scope: Scope, receive: Receive, send: Send) -> None:
        if(scope['path'] == '/graphql'):
            scope['path'] = '/graphql/'
            scope['raw_path'] = b'/graphql/'
        await self.app(scope, receive, send)
```
and
```python
app = FastAPI(redoc_url=None, docs_url=None, openapi_url=None)

app.add_middleware(GraphQLRedirect)

type_defs = load_schema_from_path("schema.graphql")
schema = make_executable_schema(
    type_defs, query, snake_case_fallback_resolvers
)

app.mount("/graphql", GraphQL(schema))
```
No more 307s!

**Comment by baebranch** [comment]
I recently ran in to this problem. It seems that the issue is with Starlette that FastAPI relies on. Starlette uses a regex pattern to match routes but the pattern used for a regular route is different from the pattern used for mounted routes so a route like `"/<path>"` doesn't actually match the regex pattern that is created. However Starletter only then checks if redirects to `"/<path>/"` work which then the mounted route's regex can match. So if your path is registered in the Sub app as `"/"` it will match `"/<path>/"` and why disabling redirects don't work and return 404 because the mounted path can't match `"/<path>"` even if your sub app registers `""` as the root.

In the end, I created a modified function of the one Starlette uses to create a regex pattern that would match ` "/<path>" and "/<path>/"` and all the other paths as it did before. So the regex patter for a mount to `"/<mount_path>"` now looks like this `'^/<mount_path>(?:/(?P<path>.*))?$'` instead of this `'^/mcp/(?P<path>.*)$'`.
